ouroboros-consensus-protocol/src/Ouroboros/Consensus/Protocol/Praos.hs

```diff
@@ -0,0 +1,684 @@
+
+deriving instance PraosCrypto c => Show (PraosValidationErr c)
+
+instance PraosCrypto c => ConsensusProtocol (Praos c) where
+  type ChainDepState _ = PraosState c
+  type IsLeader _ = PraosIsLeader c
+  type CanBeLeader _ = PraosCanBeLeader c
+  type SelectView _ = PraosChainSelectView c
+  type LedgerView _ = Views.LedgerView c
+  type ValidationErr _ = PraosValidationErr c
+  type ValidateView _ = PraosValidateView c
+
+  protocolSecurityParam = praosSecurityParam . praosParams
+
+  checkIsLeader
+    cfg
+    PraosCanBeLeader
+      { praosCanBeLeaderSignKeyVRF,
+        praosCanBeLeaderColdVerKey
+      }
+    slot
+    cs =
+      if meetsLeaderThreshold cfg lv (SL.coerceKeyRole vkhCold) y
+        then
+          Just
+            PraosIsLeader
+              { praosIsLeaderVrfRes = coerce rho
+              }
+        else Nothing
+      where
+        chainState = tickedPraosStateChainDepState cs
+        lv = getTickedPraosLedgerView $ tickedPraosStateLedgerView cs
+        eta0 = praosStateEpochNonce chainState
+        vkhCold = SL.hashKey praosCanBeLeaderColdVerKey
+        rho' = mkUnifiedVRF slot eta0
+
+        rho = VRF.evalCertified () rho' praosCanBeLeaderSignKeyVRF
+        y = vrfLeaderCheckValue (Proxy @c) $ certifiedOutput rho
+
+  -- Updating the chain dependent state for Praos.
+  --
+  -- If we are not in a new epoch, then nothing happens. If we are in a new
+  -- epoch, we do two things:
+  -- - Update the epoch nonce to the combination of the candidate nonce and the
+  --   nonce derived from the last block of the previous epoch.
+  -- - Update the "last block of previous epoch" nonce to the nonce derived from
+  --   the last applied block.
+  tickChainDepState
+    PraosConfig {praosEpochInfo}
+    (TickedPraosLedgerView lv)
+    slot
+    st =
+      TickedPraosState
+        { tickedPraosStateChainDepState = st',
+          tickedPraosStateLedgerView = TickedPraosLedgerView lv
+        }
+      where
+        newEpoch =
+          isNewEpoch
+            (History.toPureEpochInfo praosEpochInfo)
+            (praosStateLastSlot st)
+            slot
+        st' =
+          if newEpoch
+            then
+              st
+                { praosStateEpochNonce =
+                    praosStateCandidateNonce st
+                      ⭒ praosStateLastEpochBlockNonce st,
+                  praosStateLastEpochBlockNonce = praosStateLabNonce st
+                }
+            else st
+
+  -- Validate and update the chain dependent state as a result of processing a
+  -- new header.
+  --
+  -- This consists of:
+  -- - Validate the VRF checks
+  -- - Validate the KES checks
+  -- - Call 'reupdateChainDepState'
+  --
+  updateChainDepState
+    cfg@( PraosConfig
+            PraosParams {praosLeaderF}
+            _
+          )
+    b
+    slot
+    tcs = do
+      validateVRFSignature (praosStateEpochNonce cs) lv praosLeaderF b
+      validateKESSignature cfg lv (praosStateOCertCounters cs) b
```

Comment:
In addition to swapping these, I think it'd be nice to comment that the KES check are authentication checks (we are who we say we are), and the VRF is authorisation (we're the slot leader).